AUT-1: Support for forcing users to change their password on login

gitcliff opened the pull request:
PR is adding the ability to set require password change on first login via the openmrs runtime properties

ibacher commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -134,6 +134,11 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
 							session.refreshDefaultLocale(); // Refresh context locale after authentication
 							String successUrl = determineSuccessRedirectUrl(request);
+							// Check if password change on first login is required
+                            if (AuthenticationConfig.isPasswordChangeOnFirstLoginRequired() && userLogin.isFirstLogin()) {
+                                 String passwordChangeUrl = request.getContextPath() + "/passwordChange.htm";
+                                  response.sendRedirect(passwordChangeUrl);
+                            }
```
Fix the indentation here...

ibacher commented on omod/src/test/java/org/openmrs/module/authentication/web/AuthenticationFilterTest.java:
```diff
@@ -142,6 +143,29 @@ public void shouldRedirectToSuccessUrlIfAuthenticationSucceeds() throws Exceptio
+		authenticationSession.setAuthenticatedUser(null);
+		request.setMethod("GET");
+
+	    request.addParameter("redirect", "/passwordChange.htm");
+		request.addParameter("username", "admin");
+		request.addParameter("password", "adminPassword");
+        // Run the filter
+        filter.doFilter(request, response, chain);
+
+		assertThat(response.isCommitted(), equalTo(true));
+        
+        // Verify redirection to password change URL
```
Indentation

ibacher commented on api/src/main/java/org/openmrs/module/authentication/AuthenticationConfig.java:
```diff
@@ -38,6 +38,12 @@ public class AuthenticationConfig implements Serializable {
 
+    public static final String REQUIRE_PASSWORD_CHANGE_ON_FIRST_LOGIN = "authentication.requirePasswordChangeOnFirstLogin";
+
+    public static boolean isPasswordChangeOnFirstLoginRequired() {
```
Isn't this covered by `getBoolean()` below?

ibacher commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -134,6 +134,11 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
 							String successUrl = determineSuccessRedirectUrl(request);
+							// Check if password change on first login is required
+                            if (AuthenticationConfig.isPasswordChangeOnFirstLoginRequired() && userLogin.isFirstLogin()) {
+                                 String passwordChangeUrl = request.getContextPath() + "/passwordChange.htm";
+                                  response.sendRedirect(passwordChangeUrl);
```
So the page to redirect to _must_ be configurable. I would also plug this code into the `determineSuccessRedirectUrl()` function, so we only have a single part of code actually emitting a redirect.

ibacher commented on api/src/main/java/org/openmrs/module/authentication/UserLogin.java:
```diff
@@ -157,6 +158,20 @@ public synchronized void setUsername(String username) {
+    /**
+     * @return true if it's the user's first login
+     */
+    public boolean isFirstLogin() {
+        return isFirstLogin;
+    }
```
I don't see why we need these? The only thing we _should_ be checking is whether the _user_ has the user property `OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD` set to "true". That way we support both first login and additional password reset changes.

ibacher reviewed (changes requested):
Thanks @gitcliff! This is a reasonable first-pass, but see my comments. Also we should definitely have a ticket in the Authentication module Jira project for this. Could you please create one? I can fill in details.

gitcliff commented:
> Thanks @gitcliff! This is a reasonable first-pass, but see my comments. Also we should definitely have a ticket in the Authentication module Jira project for this. Could you please create one? I can fill in details.
thanks @ibacher for the review ,,,i have addressed the comments ..
I don't see the [JIRA board for this module ](https://openmrs.atlassian.net/jira/projects?page=1&sortKey=name&sortOrder=ASC) to file a ticket , should i create one ?

ibacher commented:
Created one here: https://openmrs.atlassian.net/jira/software/projects/AUT/

ibacher commented on api/src/main/java/org/openmrs/module/authentication/AuthenticationConfig.java:
```diff
@@ -38,6 +38,10 @@ public class AuthenticationConfig implements Serializable {
      */
     public static final String SCHEME_ID = "{schemeId}";
 
+    public static final String REQUIRE_PASSWORD_CHANGE_ON_FIRST_LOGIN = "authentication.requirePasswordChangeOnFirstLogin";
```
Let's rename this to `SUPPORT_FORCED_PASSWORD_CHANGE`

ibacher commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -225,6 +226,11 @@ protected String determineSuccessRedirectUrl(HttpServletRequest request) {
 		if (StringUtils.isNotBlank(redirect)) {
 			return contextualizeUrl(request, redirect);
 		}
+		// Check if password change on first login is required
+		if (AuthenticationConfig.getBoolean(AuthenticationConfig.REQUIRE_PASSWORD_CHANGE_ON_FIRST_LOGIN, false) &&
+				"true".equals(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD)) {
+			return contextualizeUrl(request, AuthenticationConfig.PASSWORD_CHANGE_URL);
+		}
```
This actually needs to come _before_ the normal redirect processing

gitcliff replied:
thanks @ibacher ,,,
kindly review

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -217,6 +218,11 @@ protected AuthenticationScheme getAuthenticationScheme() {
+		if (AuthenticationConfig.getBoolean(AuthenticationConfig.SUPPORT_FORCED_PASSWORD_CHANGE, false) &&
+				"true".equals(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD)) {
+			return contextualizeUrl(request, AuthenticationConfig.PASSWORD_CHANGE_URL);
+		}
```
@gitcliff I wouldn't think that this is something that we are just aiming to do once, as soon as the user is authenticated, and only then.  For example, if they authenticate, and we redirect them to this page, and then they just choose to enter a different url into the address bar, this wouldn't prevent them from doing that.  My assumption is that we want to force them to change their password without accessing any other pages.  So, this is really something we want to do independently of authentication.  I would think this could either be achieved via a separate servlet filter that runs after the AuthenticationFilter, or a separate check in the AuthenticationFilter that is placed at the end (eg. the code that is hit if the user is already authenticated).  Interested in @ibacher 's thoughts...

ibacher replied:
Hmmmm... That's actually a good point. Maybe having this as a second filter makes the most sense. There will need to be some additional configuration around this too... I.e., we need to white-list certain URLs (e.g., to allow the user to hit the "change password" endpoint).

gitcliff replied:
thanks @ibacher  and @mseaton

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -217,6 +218,11 @@ protected AuthenticationScheme getAuthenticationScheme() {
 	protected String determineSuccessRedirectUrl(HttpServletRequest request) {
+		// Check if password change on first login is required
+		if (AuthenticationConfig.getBoolean(AuthenticationConfig.SUPPORT_FORCED_PASSWORD_CHANGE, false) &&
+				"true".equals(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD)) {
```
This is not actually checking the user property value, but is checking the value of the constant.  You actually need to fetch the user property value with this name, not just compare the constant itself.

mseaton commented on api/src/main/java/org/openmrs/module/authentication/AuthenticationConfig.java:
```diff
@@ -70,6 +72,11 @@ public class AuthenticationConfig implements Serializable {
+    /**
+     * URLs that might need to be made accessible without go through the password change authentication.
+     */
+    public static final String PASS_WORD_CHANGE_WHITE_LIST = "authentication.passWordChangeWhiteList";
```
`PASS_WORD` -> `PASSWORD`, and `passWord` -> `password`

mseaton commented on api/src/main/java/org/openmrs/module/authentication/AuthenticationConfig.java:
```diff
@@ -234,6 +240,19 @@ public static List<String> getWhiteList() {
+    }
+
+    public static String getChangePasswordUrl() {
+        return getConfig().getProperty(PASSWORD_CHANGE_URL, "/changePassword");
```
Does this refer to an actual /changePassword page somewhere?  If we are going to have a default, we should make sure it actually works under some standard conditions (eg. with legacyui running)

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * This filter checks if an authenticated user has been flagged by the admin to change his password
+ * on first/subsequent login. It will intercept any requests made to a *.html or a *.form to force
+ * the user to change his password.
```
Why just a *.html and *.form?  I don't think this comment is accurate, is it?

gitcliff replied:
thanks  @mseaton
 had skipped removing it but removed now

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+	
+	private String excludeURL;
+	
+	private String changePasswordForm;
```
This doesn't seem right.  Why do you have this here?  Aren't you getting this from the AuthenticationConfig, which is able to be specified at runtime, not a hard-coded filter param?  Remove this.

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/AuthenticationFilter.java:
```diff
@@ -217,33 +184,23 @@ protected AuthenticationScheme getAuthenticationScheme() {
 	protected String determineSuccessRedirectUrl(HttpServletRequest request) {
+		// Check if password change on first login is required
+		if (AuthenticationConfig.getBoolean(AuthenticationConfig.SUPPORT_FORCED_PASSWORD_CHANGE, false) &&
+		    AuthenticationConfig.getBoolean(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD, false)) {
```
This is not correct.  The USER_PROPERTY_CHANGE_PASSWORD is not coming from the AuthenticationConfig, but from the user properties.

mseaton replied:
Actually, sorry - this entire set of changes should not be here at all.  This is now covered by the other filter.  You need to remove these changes to this method.

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+		AuthenticationSession session = getAuthenticationSession(request, response);
+		UserLogin userLogin = session.getUserLogin();
+		String changePasswordProperty = userLogin.getUser().getUserProperties()
+				.get(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD);
```
You should be able to use `userLogin.getUser().getUserProperty(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD)`

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+			throws IOException, ServletException {
+		HttpServletRequest request = (HttpServletRequest) servletRequest;
+		HttpServletResponse response = (HttpServletResponse) servletResponse;
+
```
Previously, you had a check for `(AuthenticationConfig.getBoolean(AuthenticationConfig.SUPPORT_FORCED_PASSWORD_CHANGE, false)`.  Are you no longer supporting this?  I would expect to see that in this filter.  And I'd also recommend checking this once and saving it so that you can make this filter as efficient as possible for every request.

gitcliff replied:
thanks @mseaton 
i have added the check and cached it

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+		UserLogin userLogin = session.getUserLogin();
+		String changePasswordProperty = userLogin.getUser().getUserProperties()
+				.get(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD);
+		Boolean changePasswordFlag = AuthenticationConfig.getBoolean(changePasswordProperty, false);
```
This is not correct.  You don't use the AuthenticationConfig to do this, you would use something like `BooleanUtils.isTrue()`

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,80 @@
+				.get(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD);
+		Boolean changePasswordFlag = AuthenticationConfig.getBoolean(changePasswordProperty, false);
+		if (userLogin.isUserAuthenticated() && BooleanUtils.isTrue(changePasswordFlag)) {
+			String changePasswordUrl = AuthenticationConfig.getChangePasswordUrl();
```
This seems like something that should be evaluated once and cached, not evaluated and checked per request.

mseaton commented on api/src/main/java/org/openmrs/module/authentication/AuthenticationConfig.java:
```diff
@@ -82,6 +91,7 @@ public class AuthenticationConfig implements Serializable {
 
     private static Properties config;
 
+
```
Remove this newline

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,91 @@
+	public void init(FilterConfig config) throws ServletException {
+		this.config = config;
+		this.changePasswordUrl = AuthenticationConfig.getChangePasswordUrl();
+		this.supportForcedPasswordChange = BooleanUtils.isTrue(AuthenticationConfig.getBoolean
```
The BooleanUtils.isTrue here is unnecessary.  Just use the AuthenticationConfig method

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,91 @@
+		HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+		if (WebUtil.isWhiteListed(request, AuthenticationConfig.getPasswordChangeWhiteList()) 
+		    || !supportForcedPasswordChange) {
```
I would evaluate `supportForcedPasswordChange` first, as it is faster and if it is true then it won't bother evaluating the next.  Why aren't you also caching the whitelist value at startup?

gitcliff replied:
yeah, thanks @mseaton ,, cached the whitelist value at startup

mseaton commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,91 @@
+	 * @return the AuthenticationSession associated with this HttpServletRequest
+	 */
+	protected AuthenticationSession getAuthenticationSession(HttpServletRequest request, 
+	  HttpServletResponse response) {
```
Ugly

mseaton reviewed (changes requested):
Looking better.  Please re-request review via the github PR tools when you are ready for this to be looked at again.

gitcliff commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -0,0 +1,92 @@
+		UserLogin userLogin = session.getUserLogin();
+		String changePasswordProperty = userLogin.getUser().
+		  getUserProperty(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD);
+		Boolean changePasswordFlag = BooleanUtils.isTrue(Boolean.valueOf(changePasswordProperty));
```
thanks @ibacher fixed it

ibacher commented on omod/src/main/java/org/openmrs/module/authentication/web/ForcePasswordChangeFilter.java:
```diff
@@ -64,14 +67,9 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
 		  getUserProperty(OpenmrsConstants.USER_PROPERTY_CHANGE_PASSWORD);
-		Boolean changePasswordFlag = BooleanUtils.isTrue(Boolean.valueOf(changePasswordProperty));
+		boolean changePasswordFlag = Boolean.parseBoolean(changePasswordProperty);
 		if (userLogin.isUserAuthenticated() && BooleanUtils.isTrue(changePasswordFlag)) {
```
So, as I noted, you can get rid of the `BooleanUtils.isTrue()` call on this line too.
